### Layer Initialization and Transfer Learning

Data normalization scales input features for stability; similarly, the weights of a linear layer are also initialized to small values. This is known as ***Layer Initialization***.

> Let's create a small linear layer and check its weight range.

In [2]:
import torch.nn as nn

layer = nn.Linear(64, 128)
print(layer.weight.min(), layer.weight.max())

tensor(-0.1250, grad_fn=<MinBackward1>) tensor(0.1249, grad_fn=<MaxBackward1>)


We observe that the weights are between ***-0.125*** and ***0.125***. 

Why is it important? the output of a neuron in a linear layer is a weighted sum of inputs from the previous layer. Keeping both the input data and layer weights small ensures stable outputs, preventing extreme values that could slow training. Layers can be intialized in different ways and it remains an active area of research.

---

PyTorch provides an easy way to initialize layers weights with the ***nn.init*** module. For example, here we initialize a linear layer with an uniform distribution. As you can see, the weights values now range from 0 to 1.

In [3]:
import torch.nn as nn

layer = nn.Linear(64, 128)
nn.init.uniform_(layer.weight)

print(layer.weight.min(), layer.weight.max())

tensor(0.0002, grad_fn=<MinBackward1>) tensor(0.9999, grad_fn=<MaxBackward1>)


### Transfer Learning

In practice, engineers are rarely training a model from randomly initialized weights. Instead, they rely on a concept called ***transfer learning***. 

Transfer Learning takes a model that was trained on a first task and reuses it for a second task. For example, we trained a model on US data of salaries in Europe. Instead of traininga model using randomly initialized weights, we can load the weights from the first model and use them as a starting point to train on this new dataset. 

Saving and loading weights can be done using the ***torch.save*** and the ***torch.load*** functions. These functions work on any type of PyTorch objects.

In [ ]:
import torch

layer = nn.Linear(64, 128)
torch.save(layer, 'layer.pth')

new_layer = torch.load('layer.pth')

### Fine Tuning

Sometimes, the second task is similaer to the first task and we want to perform a specific type of transfer learning, called ***fine-tuning***.

In this case, we load weights from a previously trained model, but train the model with a smaller learning rate. We can even train part of a network, if we decide some of the network layers do not need to be trained and choose to freeze them.

A rule of thumb is to ***freeze*** early layers of the netwrok and ***fine-tune*** layers closer to the output layer. This can be achieved by setting each parameter's ***<requires_grad_attribute>*** to ***<False>***. 

Here, we use the model's named_parameters() method, which returns the name and the parameter itself. We set requires_grad of the first layer's weight to False.

In [2]:
import torch.nn as nn

model = nn.Sequential(nn.Linear(64, 128),
                      nn.Linear(128, 256))

for name, param in model.named_parameters():
    if name == '0.weight':
        param.requires_grad = False